webbpsf/roman.py

```diff
@@ -331,171 +332,196 @@ def __init__(self):
-
-        if 'AUTO' == self.pupil_mask:
-            if self._currently_masked:
-                self.pupil = self._masked_pupil_path
-            else:
-                self.pupil = self._unmasked_pupil_path
-        elif self.pupil_mask in ["FULL_MASK", "COLD_PUPIL"]:
-            self.pupil = self._masked_pupil_path
-        elif self.pupil_mask in ["RIM_MASK", "UNMASKED"]:
-            self.pupil = self._unmasked_pupil_path
-        else:
-            raise ValueError("Pupil mask setting is not valid or empty.")
+        Undoes the effects of lock_pupil() and resets WFIPupilController
+        to its default state of updating the pupil whenever a detector
+        or filter is changed in the WFI class.
+        """
+        self._auto_pupil = True
 
-    def validate_pupil(self, filter, **kwargs):
-        """Validates that the WFI is configured sensibly
+    def lock_pupil_mask(self, pupil_mask):
+        """
+        Prevents the WFIPupilController class from dynamically updating
+        the pupil mask on any changes to the filter selected in the WFI
+        class. Instead, the pupil mask remains locked on whichever
+        `pupil_mask` was provided to this method.
 
-        This mainly consists of selecting the masked or unmasked pupil
-        appropriately based on the wavelengths requested.
+        CAUTION: This is non-standard usage of the WFI class and may
+        lead to unexpected behavior.
+
+        Parameters
+        ----------
+        filter : string
+            See WFI.pupil_mask_list for a list of valid pupil masks.
         """
-        if self.auto_pupil:
-            if filter in self._masked_filters:
-                # use masked pupil optic
-                self.pupil = self._masked_pupil_path
-                _log.info("Using the masked WFI pupil shape based on filter requested")
-            else:
-                # use unmasked pupil optic
-                self.pupil = self._unmasked_pupil_path
-                _log.info("Using the unmasked WFI pupil shape based on filter requested")
+        if pupil_mask not in self.pupil_file_formatters.keys():
+            raise Exception('invalid pupil mask')
+        elif not self._auto_pupil:
+            raise Exception('Pupil is locked. Unlock pupil before locking pupil mask.')
         else:
-            # If the user has set the pupil to a custom value, let them worry about the
-            # correct shape it should have
-            pass
+            self._pupil_mask = pupil_mask
+            self._auto_pupil_mask = False
+
+    def unlock_pupil_mask(self):
+        """
+        Undoes the effects of lock_pupil_mask() and resets
+        WFIPupilController to its default state of updating the pupil
+        mask whenever filter is changed in the WFI class.
 
-    def remove_pupil_mask_override(self):
-        _log.info("Removing custom pupil mask")
-        self.pupil_mask = 'AUTO'
+        UNRESOLVED: how do we know which mask to reset to?
+        """
```

Comment:
You're right in saying that `WFIPupilController` has no knowledge of the current pupil or pupil mask value and that the actual assignment of the `pupil_mask` attribute happens in `WFI`. This note was a reminder to re-check the WFI version of `unlock_pupil_mask()` since it calls this one.